src/media/mod.rs

```diff
@@ -0,0 +1,84 @@
+//!Media module
+//!
+//!Provides functionality to upload images, GIFs and videos using Twitter Media API
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use error;
+use error::Error::InvalidResponse;
+
+use rustc_serialize::{
+    json,
+    base64
+};
+use self::base64::{ToBase64};
+
+use common::*;
+use links;
+use auth;
+
+///Represents media file that is uploaded on twitter.
+pub struct Media {
+    ///ID that can be used in API calls (e.g. attach to tweet).
+    pub id: u64,
+    ///Number of second the media can be used in other API calls.
+    pub expires_after: u64,
+}
+
+impl FromJson for Media {
+    fn from_json(input: &json::Json) -> Result<Self, error::Error> {
+        if !input.is_object() {
+            return Err(InvalidResponse("Tweet received json that wasn't an object", Some(input.to_string())));
+        }
+
+        field_present!(input, media_id);
+        field_present!(input, expires_after_secs);
+
+        Ok(Media {
+            id: try!(field(input, "media_id")),
+            expires_after: try!(field(input, "expires_after_secs"))
+        })
+    }
+}
+
+///Uploads image using old twitter API.
+///
+///The image should be raw binary with content of file.
+pub fn upload_image<'a, I: Into<Cow<'a, [u8]>>>(image: I, token: &auth::Token, handle: &'a Handle) -> FutureResponse<'a, Media> {
+    let mut params = HashMap::new();
+
+    let config = base64::Config {
+        char_set: base64::CharacterSet::Standard,
+        newline: base64::Newline::LF,
+        pad: true,
+        line_length: None,
+    };
+
+    add_param(&mut params, "media_data", image.into().to_base64(config));
```

Comment:
Come to think of it, now that we're making a new String for it, we probably don't have to take a Cow for the bytes. We could get away with just taking a `&[u8]` (we can leave the lifetime inferred and the compiler will make a separate one from the Handle) since we don't need to own it for the upload itself.

Reply:
Yeah, that's true